languages/python/sqlalchemy-oso/sqlalchemy_oso/session.py

```diff
@@ -219,3 +223,53 @@ class AuthorizedSession(AuthorizedSessionBase, Session):
     """
 
     pass
+
+
+if AT_LEAST_SQLALCHEMY_VERSION_1_4:
+    # TODO(gj): remove type ignore once we upgrade to 1.4-aware MyPy types.
+    from sqlalchemy.orm import with_loader_criteria  # type: ignore
+
+    @event.listens_for(Session, "do_orm_execute")
+    def do_orm_execute(execute_state):
+        if not execute_state.is_select:
+            return
+
+        session = execute_state.session
+
+        if not isinstance(session, AuthorizedSessionBase):
+            return
+        assert isinstance(session, Session)
+
+        oso: Oso = session.oso_context["oso"]
+        user = session.oso_context["user"]
+        action = session.oso_context["action"]
+
+        def get_entities(x):
+            def _get_entities(x):
+                try:
+                    return set(
+                        e
+                        for e in (cd["entity"] for cd in x.column_descriptions)
+                        if e is not None
+                    )
+                except AttributeError:
+                    return set()
+
+            # TODO(gj): currently walking way more than we have to. Probably
+            # some points in the tree where we can safely call it good for that
+            # branch and continue on to more fruitful pastures.
+            def _walk_children(x):
+                return x.get_children()
+
+            entities = _get_entities(x)
+
+            for child in _walk_children(x):
+                entities |= get_entities(child)
+
+            return entities
```

Comment:
I'm sure there's a way to Python better than I'm doing it here, but this seems to work.

The walk is to ensure we pick up entities in subqueries. For example, there's no top-level `User` entity in the following query, but there is one in the subquery, and by walking the whole query we find the nested `User` entity and then apply authorization to User model accesses for the entire query:

```sql
SELECT count(*) AS count_1
FROM (SELECT users.id AS users_id, users.username AS users_username, users.is_moderator AS users_is_moderator, users.is_banned AS users_is_banned, users.tag_name AS users_tag_name
FROM users) AS anon_1
```